Regression: Index containing class names that start with $ no longer have a consistent DotName.

CC @MikeEdgar 
This appears to be a regression with the parsing of DotNames caused by https://github.com/wildfly/jandex/commit/3779e75f7369ad4849f271b378fa2370f3d0e0f3

Consider the following test condition for the org.jboss.cdi.tck.tests.definition.name.$Dollar class:

``` java
Class<?> targetClass = $Dollar.class;
Indexer indexer = new Indexer();
try (InputStream stream = targetClass.getClassLoader().getResourceAsStream(targetClass.getName().replace('.', '/') + ".class")) {
    indexer.index(stream);
    Index index = indexer.complete();
    DotName name = DotName.createSimple(targetClass.getName());
    Assert.assertNotNull(index.getClassByName(name));
}
```

This assertion passes in 2.3.1, but fails in 2.4.0.
If I revert the above commit, the above assertion succeeds.

This is somewhat critical, as it causes a regression in the Jakarta TCK for CDI running against WildFly + Jandex 2.4.0 for the following test:
https://github.com/eclipse-ee4j/cdi-tck/blob/3.0/impl/src/main/java/org/jboss/cdi/tck/tests/definition/name/NameDefinitionTest.java#L109

```
[ERROR] org.jboss.cdi.tck.tests.definition.name.NameDefinitionTest.testNameStartingWithDollarCharacter  Time elapsed: 0.049 s  <<< FAILURE!
jakarta.enterprise.inject.UnsatisfiedResolutionException: Unable to resolve any beans of class org.jboss.cdi.tck.tests.definition.name.$Dollar
	at deployment.9d9c1e656827b7ae88a4f45c576e6894daaaf63.war//org.jboss.cdi.tck.tests.definition.name.NameDefinitionTest.testNameStartingWithDollarCharacter(NameDefinitionTest.java:110)
```

... and thus prevents WildFly from upgrading any of the smallrye components that require Jandex 2.4.0 (e.g. smallrye-open-api).

--- a/src/main/java/org/jboss/jandex/NameTable.java
+++ b/src/main/java/org/jboss/jandex/NameTable.java
@@ -45,7 +45,7 @@ class NameTable {
         if (result != null)
             return result;
 
-        int loc = lastIndexOf(name, delim, '$');
+        int loc = lastIndexOf(name, delim);
         String local = intern(name.substring(loc + 1));
         DotName prefix = loc < 1 ? null : convertToName(intern(name.substring(0, loc)), delim);
         result = new DotName(prefix, local, true, loc > 0 && name.charAt(loc) == '$');
@@ -55,23 +55,21 @@ class NameTable {
         return result;
     }
 
-    private int lastIndexOf(String name, char delim1, char delim2) {
+    private int lastIndexOf(String name, char delim) {
         // Begin at second last position to avoid empty local name
         int pos = name.length() - 1;
         while (--pos >= 0) {
             char c = name.charAt(pos);
-
-            if (c == delim1) {
+            if (c == delim || c == '$') {
                 break;
             }
+        }
 
-            /*
-             * Only split on the inner class delimiter when it will not
-             * result in an empty string for the containing top-level class.
-             */
-            if (c == delim2 && pos > 0 && name.charAt(pos - 1) != delim1) {
-                break;
-            }
+        // avoid splitting on '$' if previous char is a delimiter or the '$'
+        // is in position 0, because subsequent split would produce an empty
+        // local name
+        if (pos >=0 && name.charAt(pos) == '$' && (pos == 0 || name.charAt(pos - 1) == delim)) {
+            pos--;
         }
 
         return pos;

--- a/src/main/java/org/jboss/jandex/NameTable.java
+++ b/src/main/java/org/jboss/jandex/NameTable.java
@@ -60,7 +60,16 @@ class NameTable {
         int pos = name.length() - 1;
         while (--pos >= 0) {
             char c = name.charAt(pos);
-            if (c == delim1 || c == delim2) {
+
+            if (c == delim1) {
+                break;
+            }
+
+            /*
+             * Only split on the inner class delimiter when it will not
+             * result in an empty string for the containing top-level class.
+             */
+            if (c == delim2 && pos > 0 && name.charAt(pos - 1) != delim1) {
                 break;
             }
         }

--- a/src/test/java/$delimiters$/test/$SurroundedByDelimiters$.java
+++ b/src/test/java/$delimiters$/test/$SurroundedByDelimiters$.java
@@ -1,0 +1,5 @@
+package $delimiters$.test;
+
+public class $SurroundedByDelimiters$ {
+
+}

--- a/src/test/java/$pkg/test/$LeadingDelimiter.java
+++ b/src/test/java/$pkg/test/$LeadingDelimiter.java
@@ -1,0 +1,5 @@
+package $pkg.test;
+
+public class $LeadingDelimiter {
+
+}

--- a/src/test/java/org/jboss/jandex/test/DotNameTestCase.java
+++ b/src/test/java/org/jboss/jandex/test/DotNameTestCase.java
@@ -237,14 +237,29 @@ public class DotNameTestCase {
 
     @Test
     public void testLeadingInnerClassDelimiterOnClass() throws IOException {
-        DotName $pkg = DotName.createComponentized(null, "$pkg");
-        DotName test = DotName.createComponentized($pkg, "test");
+        DotName pkg = DotName.createComponentized(null, "$pkg");
+        DotName test = DotName.createComponentized(pkg, "test");
         DotName testName = DotName.createComponentized(test, $LeadingDelimiter.class.getSimpleName());
 
         Index index = Index.of($LeadingDelimiter.class);
         assertEquals(testName, index.getKnownClasses().iterator().next().name());
         assertNotNull(index.getClassByName(DotName.createSimple($LeadingDelimiter.class.getName())));
         assertNotNull(index.getClassByName(testName));
+    }
+
+    @Test
+    public void testClassNameWithDelimitersFirstAndLast() throws IOException {
+        DotName pkg = DotName.createComponentized(null, "$delimiters$");
+        DotName test = DotName.createComponentized(pkg, "test");
+        DotName testName = DotName.createComponentized(test, $delimiters$.test.$SurroundedByDelimiters$.class.getSimpleName());
+        DotName testNameSimple = DotName.createSimple($delimiters$.test.$SurroundedByDelimiters$.class.getName());
+
+        Index index = Index.of($delimiters$.test.$SurroundedByDelimiters$.class);
+        DotName indexedName = index.getKnownClasses().iterator().next().name();
+        assertEquals(testName, indexedName);
+        assertNotNull(index.getClassByName(testNameSimple));
+        assertNotNull(index.getClassByName(testName));
+        assertEquals("$delimiters$.test.$SurroundedByDelimiters$", indexedName.toString());
     }
 
     private static DotName createRandomDotName() {

--- a/src/test/java/org/jboss/jandex/test/DotNameTestCase.java
+++ b/src/test/java/org/jboss/jandex/test/DotNameTestCase.java
@@ -20,8 +20,8 @@ package org.jboss.jandex.test;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
 
 import java.io.IOException;
 import java.util.Iterator;
@@ -32,6 +32,8 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.Index;
 import org.junit.Assert;
 import org.junit.Test;
+
+import $pkg.test.$LeadingDelimiter;
 
 /**
  * Since DotName is often used as a key in collections and implements Comparable,
@@ -231,6 +233,18 @@ public class DotNameTestCase {
 
     public static class Test$ {
 
+    }
+
+    @Test
+    public void testLeadingInnerClassDelimiterOnClass() throws IOException {
+        DotName $pkg = DotName.createComponentized(null, "$pkg");
+        DotName test = DotName.createComponentized($pkg, "test");
+        DotName testName = DotName.createComponentized(test, $LeadingDelimiter.class.getSimpleName());
+
+        Index index = Index.of($LeadingDelimiter.class);
+        assertEquals(testName, index.getKnownClasses().iterator().next().name());
+        assertNotNull(index.getClassByName(DotName.createSimple($LeadingDelimiter.class.getName())));
+        assertNotNull(index.getClassByName(testName));
     }
 
     private static DotName createRandomDotName() {

